### PR TITLE
add sampling support

### LINKED ISSUE #156: Query turn on/off Sampling mode
Dashbase query use keyword turn on/off sampling mode
e.g.
```
!~~        
# trun off sampling
~~ 
# trun on sampling

_mode:sampling
# trun on sampling
_mode:!sampling 
# trun off sampling

```
if timeRange bigger than 15m, auto use sampling, and could use keyword to trun off it.

---
#155 
#83

### PR DESCRIPTION
**Pull request by WPH95**
resolved #156

### CODE REVIEW

File: src/main/java/io/dashbase/es/SearchFactory.java [hunk 1]
```diff
@@ -38,6 +39,14 @@ public QueryConverter getQueryConverter() {
         return queryConverter;
     }
 
+    static final long APPROX_THRESHOLD_IN_SEC = 15 * 60; // fifteen min
+
+    public Optional<Boolean> isSampling = Optional.empty();
+
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/es/SearchFactory.java]: 'Optional<Boolean>' used as type for field 'isSampling'
Reports any uses of java.util.Optional<T>, java.util.OptionalDouble, java.util.OptionalInt, java.util.OptionalLong or com.google.common.base.Optional as the type for a field or a parameter. Optional was designed to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result". Using a field with type java.util.Optional is also problematic if the class needs to be Serializable, which java.util.Optional is not.

Comment (lfkdsk) [reply src/main/java/io/dashbase/es/SearchFactory.java]: 这里的是指，Optional 的用法是指在接受返回参数的时候的一种优雅的判空方法。而不是说把这东西当成一个全局变量的类型来用。

File: src/main/java/io/dashbase/es/SearchFactory.java [hunk 2]
```diff
@@ -38,6 +39,14 @@ public QueryConverter getQueryConverter() {
         return queryConverter;
     }
 
+    static final long APPROX_THRESHOLD_IN_SEC = 15 * 60; // fifteen min
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/es/SearchFactory.java]: private

File: src/main/java/io/dashbase/es/SearchFactory.java [hunk 3]
```diff
@@ -38,6 +39,14 @@ public QueryConverter getQueryConverter() {
...
+
+    public Optional<Boolean> isSampling = Optional.empty();
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/es/SearchFactory.java]: 能 private 用 getter 就别 public 裸。

File: src/main/java/io/dashbase/utils/query/decode.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import rapid.api.query.Query;
+import rapid.api.query.StringQuery;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+public class decode {
+
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/utils/query/decode.java]: 驼峰类名

File: src/main/java/io/dashbase/utils/query/decode.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    public static Optional<Boolean> isSamplingModeToken(String token) {
+        switch (token) {
+            case "~~":
+            case "_mode:sampling":
+                return Optional.of(true);
+            case "!~~":
+            case "_mode:!sampling":
+                return Optional.of(false);
+            default:
+                return Optional.empty();
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/utils/query/decode.java]: 不存在的情况是不是应该有 exception ？

File: src/main/java/io/dashbase/utils/query/decode.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+
+    public static Optional<Boolean> isSamplingQuery(Conjunction query, Boolean clean) {
+        Optional<Boolean> result = Optional.empty();
+        for (Query subQuery : query.subQueries) {
+            if (subQuery instanceof StringQuery) {
+                result = isSamplingQuery((StringQuery) subQuery, clean);
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/utils/query/decode.java]: Query 要是能靠 Field 变量判断类型就好了，使用 instanceOf 会影响效率。

File: src/main/java/io/dashbase/utils/query/decode.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                    query.subQueries.remove(subQuery);
+            }
+        }
+        return result;
+
+    }
+
+    public static Optional<Boolean> isSamplingQuery(StringQuery query, Boolean clean) {
+        List<String> token = new LinkedList<>(Arrays.asList(split(query.queryStr)));
+        Optional<Boolean> result = Optional.empty();
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/utils/query/decode.java]: asList 本身就创建了一个新的 List ，不需要 protected-copy 了。

File: src/main/java/io/dashbase/es/SearchFactory.java [hunk 8]
```diff
@@ -139,10 +157,15 @@ private void decodeQuery() {
 
         // clean up empty empty subQuery list(subquery:[]).
         dashbaseQueries.removeIf(e -> (e instanceof Conjunction && ((Conjunction) e).subQueries.size() == 0));
+
+        if (dashbaseQueries.size() > 1) {
+            request.query = new Conjunction(dashbaseQueries);
```
Comment (lfkdsk) [inline src/main/java/io/dashbase/es/SearchFactory.java]: `!dashBaseQueries.isEmpty()`

Comment (WPH95) [reply src/main/java/io/dashbase/es/SearchFactory.java]: >1 
不是 >0

File: src/main/java/io/dashbase/utils/query/decode.java (hunk 6 above)
Comment (WPH95) [reply src/main/java/io/dashbase/utils/query/decode.java]: 之后搞一下吧

**Pull request merged**